java/arcs/core/util/Parser.kt

```diff
@@ -0,0 +1,278 @@
+}
+
+/** Chop off the consumed part of the string. */
+fun String.advance(str: String) = this.substring(str.length)
+
+/** A parser that consumes a prefix of a string. */
+class StringToken(val token: String) : Parser<String>() {
+    override fun invoke(string: String): ParseResult<String> = when {
+        string.startsWith(token) -> Success(token, string.advance(token))
+        else -> Failure
+    }
+}
+
+/**
+ * A parser that consumes a regex prefix (anchored at ^) of a String, and returns the
+ * first matchgroup.
+ */
+class RegexToken(val regexToken: String) : Parser<String>() {
+    override fun invoke(string: String): ParseResult<String> =
+        Regex("^${regexToken}").find(string, 0)?.let { it ->
+            Success(it.groupValues[1], string.advance(it.groupValues[0]))
+        } ?: Failure
+}
+
+/**
+ * A parser that invokes another parser, and if it success, it returns, otherwise is succeeds with
+ * null.
+ */
+class Optional<T>(val parser: Parser<T>) : Parser<T?>() {
+    override fun invoke(string: String): ParseResult<T?> = parser(string).orElse {
+        Success<T?>(null, string)
+    }
+}
+
+/**
+ * A parser that combines two parsers in sequence, if either one fails, the parse fails,
+ * otherwise the combined results are returned as a [Pair<T, S].
+ */
+class Seq<T, S>(val left: Parser<T>, val right: Parser<S>) : Parser<Pair<T, S>>() {
+    override fun invoke(string: String) = left(string).flatMap { v1, r ->
+        right(r).flatMap { v2, r2 ->
+            Success(Pair(v1, v2), r2)
+        }
+    }
+}
+
+/** A parser that combines two parsers by returning the value of the first one that succeeds. */
+class Par<T>(val left: Parser<T>, val right: Parser<T>) : Parser<T>() {
+    override fun invoke(string: String): ParseResult<T> = left(string).orElse {
+        right(string)
+    }
+}
+
+/**
+ * A parser that invokes another parser zero or more times until it fails, returning a list
+ * of success values. If no parse succeeds, it returns an empty list.
+ */
+class Many<T>(val parser: Parser<T>) : Parser<List<T>>() {
+    override fun invoke(string: String): ParseResult<List<T>> {
+        val result = mutableListOf<T>()
+        var parseResult = parser(string)
+        var lastSuccess = parseResult
+        while (parseResult is Success<T>) {
+            result.add(parseResult.value)
+            lastSuccess = parseResult
+            parseResult = parser(parseResult.rest)
+        }
+        return Success(result, (lastSuccess as? Success<T>)?.rest ?: string)
+    }
+}
+
+/** A parser which converts the return value of a parser into another value. */
+class Transform<T, R>(val parser: Parser<T>, val transform: (T) -> R) : Parser<R>() {
+    override fun invoke(string: String): ParseResult<R> = parser(string).flatMap { v, r ->
+        Success(transform(v), r)
+    }
+}
+
+/** A parser used to refer to parsers that haven't been constructed yet. */
+class LazyParser<T>(val parser: () -> Parser<T>) : Parser<T>() {
```

Comment:
There is some asymmetry in naming. Sometimes you use abbreviations (e.g. `Seq` and `Par`), sometimes single words (`Transform`, `Many`) and sometimes you use word + Parser (`LazyParser`, `IgnoreParser`). I presume you didn't use some names (e.g. `Sequence`, `Ignore`) to avoid clashing with Kotlin stdlib.

I am wondering whether we should settle on longer names given the recommended API is to use helper functions and operators?

Reply:
I changed Par to AnyOf, Seq to PairOf,  and TriSeq to TripleOf. I'll add Parser suffix to everything since it is expected everything will be done with extensions/helpers.